compiler/modules/front/codeGen.bal

```diff
@@ -1216,6 +1215,109 @@ function codeGenExpr(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:Ex
+    if operand is bir:IntOperand {
+        return operand;
+    }
+    return cx.semanticErr("expected an int operand", subExprRange);
+}
+
+function exprRange(s:Expr expr) returns Range {
+    return { startPos: expr.startPos, endPos: expr.endPos };
+}
+
+function codeGenNilLiftResult(CodeGenContext cx, ExprEffect nonNilEffect, bir:BasicBlock? ifNilBlock, Position pos) returns ExprEffect {
+    if ifNilBlock == () {
+        return nonNilEffect;
+    }
+    else {
+        bir:BasicBlock block = cx.createBasicBlock();
+        bir:Register nonNilResult = <bir:Register> nonNilEffect.result;
+        bir:BasicBlock nonNilBlock = nonNilEffect.block;
+
+        bir:Register result = cx.createTmpRegister(t:union(nonNilResult.semType, t:NIL));
+        bir:AssignInsn nilAssign = { result, operand: (), pos };
+        ifNilBlock.insns.push(nilAssign);
+        bir:BranchInsn branchInsn = { dest: block.label, pos };
+        ifNilBlock.insns.push(branchInsn);
+
+        bir:AssignInsn valAssign = { result, operand: nonNilResult, pos };
+        nonNilBlock.insns.push(valAssign);
+        nonNilBlock.insns.push(branchInsn);
+        return { result, block };
+    }
+}
+
+type NilLiftResult record {|
+    bir:Operand[] operands;
+    bir:BasicBlock nextBlock;
+    bir:BasicBlock? ifNilBlock;
+|};
+
+type BinaryNilLiftResult record {|
+    bir:Operand lhs;
+    bir:Operand rhs;
+    bir:BasicBlock nextBlock;
+    bir:BasicBlock? ifNilBlock = ();
+|};
+
+type UnaryNilLiftResult record {|
+    bir:Operand operand;
+    bir:BasicBlock nextBlock;
+    bir:BasicBlock? ifNilBlock = ();
+|};
+
+function codeGenBinaryNilLift(CodeGenContext cx, Environment env, s:Expr left, s:Expr right, bir:BasicBlock bb, Position pos) returns BinaryNilLiftResult|CodeGenError {
+    var { operands, nextBlock, ifNilBlock } = check codeGenNilLift(cx, env, [left, right], bb, pos);
+    return { lhs: operands[0], rhs: operands[1], nextBlock, ifNilBlock };
+}
+
+function codeGenUnaryNilLift(CodeGenContext cx, Environment env, s:Expr operand, bir:BasicBlock bb, Position pos) returns UnaryNilLiftResult|CodeGenError {
+    var { operands, nextBlock, ifNilBlock } = check codeGenNilLift(cx, env, [operand], bb, pos);
+    return { operand: operands[0], nextBlock, ifNilBlock };
+}
+
+function codeGenNilLift(CodeGenContext cx, Environment env, s:Expr[] operands, bir:BasicBlock bb, Position pos) returns NilLiftResult|CodeGenError {
+    bir:BasicBlock? ifNilBlock = ();
+    bir:BasicBlock nextBlock;
+    bir:BasicBlock currentBlock = bb;
+    bir:Operand[] newOperands = [];
+    foreach int i in 0 ..< operands.length() {
+        s:Expr operandExpr = operands[i];
+        var { result: operand, block } = check codeGenExpr(cx, currentBlock, env, operandExpr);
+        currentBlock = block;
+        if operand is bir:Register && t:containsNil(operand.semType) && t:singleNumericType(operand.semType) != () {
```

Comment:
What is the call to singleNumericType doing here?

Reply:
I added it to make sure only number types will be lifted. But removing doesn't seems to have an effect, so that check is removed.